### PR TITLE
Unit tests and fixes to Matchers

### PR DESCRIPTION
**Pull request by theJasonFan**
**README changes**
- added: example with equal best and second best matches
- fix: README example 3 w/ best and second best match distances 0 and 2, respectively
- fix: Whitespace and bullets in README

**Fixes**
- Refactored `should_reject_delta` so all logic is in one place
- The condition for rejection should have been `delta <= min_delta`. In English, we accept matches if delta is strictly greater than `delta`. An illustrative example is with `min_delta == 0`, the most permissive setting. When `min_delta == 0`, we accept all deltas > `0`. Note that `delta == 0` iff there are multiple best matches, and that `delta == usize::MAX` when there is only one match and that match is unique (see `find_indepedently`).

**Added basic unit tests in `matcher.rs`**
- New hamming dist test with non-zero `free_ns`
- Simple test example with `min_delta == 1`, `max_mismatches == 2`, with expected match and no matches for both matchers
- Programatic test generation for simple examples. Readme examples can be generated and tested with simple function call.

**Tests**
Run...
- Simple matcher tests with `cargo test test_matches`
- Readme example tests with `cargo test test_readme_examples`

### CODE REVIEW

**Comment by theJasonFan** [comment]
In 1adf178

**README changes**
- added: example with equal best and second best matches
- fix: README example 3 w/ best and second best match distances 0 and 2, respectively
- fix: Whitespace and bullets in README

**Added basic unit tests in `matcher.rs`**
- New hamming dist test with non-zero `free_ns`
- Simple test example with `min_delta == 1`, `max_mismatches == 2`, with expected match and no matches for both matchers
- Programatic test generation for simple examples. Readme examples can be generated and tested with simple function call.

**Tests**
Run...
- Simple matcher tests with `cargo test test_matches`
- Readme example tests with `test_readme_examples`

Current test results:

```
test matcher::test::test_hamming_dist_free_ns ... ok

test matcher::test_matches::test_find_ind_match ... ok
test matcher::test_matches::test_precompute_match ... ok
test matcher::test_matches::test_precompute_no_match ... ok
test matcher::test_matches::test_find_ind_no_match ... FAILED

test matcher::test_readme_examples::ex2 ... ok
test matcher::test_readme_examples::ex3 ... ok
test matcher::test_readme_examples::ex5 ... ok

File: src/lib/matcher.rs [hunk 1]
```diff
@@ -184,13 +192,15 @@ impl<'a> PreComputeMatcher<'a> {
...
         min_mismatches: usize,
         max_mismatches: usize,
-    ) -> impl Iterator<Item = (Vec<u8>, usize)> + '_ {
+    ) -> impl Iterator<Item = (Vec<u8>, usize)> {
         let free_ns = 0;
+        // xs.iter().combinations(n) returns [] if n > xs.len();
+        let max_mismatches = max_mismatches.min(barcode.len());
         (0..barcode.len())
-            .combinations(max_mismatches)
+            .combinations(max_mismatches) // positions to insert mismatches
```
Comment (nh13) [inline src/lib/matcher.rs:203]:
```suggestion
            .combinations(max_mismatches) // generate positions to insert mismatches
```

File: src/lib/matcher.rs [hunk 2]
```diff
@@ -239,7 +249,7 @@ impl<'a> PreComputeMatcher<'a> {
                 let best = best_hits[0];
                 if matches
                     .iter()
-                    .filter(|m| m.hamming_dist <= best.hamming_dist + min_delta)
+                    .filter(|m| should_reject_delta(m.hamming_dist - best.hamming_dist, min_delta))
```
Comment (nh13) [inline src/lib/matcher.rs:252]: 👍 👍 👍

File: src/lib/matcher.rs [hunk 3]
```diff
@@ -276,7 +293,7 @@ fn find_independently(
     }
     if best_sample == samples.len()
         || best_dist > max_mismatches
-        || next_best_dist - best_dist < min_delta
+        || should_reject_delta(next_best_dist - best_dist, min_delta)
```
Comment (nh13) [inline src/lib/matcher.rs:296]: 👍 👍 👍

**Review by nh13 — APPROVED** [review]
This LGMT!  Can you please fixup and resolve comments, then feel free to merge!

File: README.md [hunk 4]
```diff
@@ -86,12 +86,13 @@ Note: the allowed mismatches is not used when determining the next-best matching
...
-
+2. If b1 matches with 1 mismatch, and b2 matches with 3 mismatches, then the delta between the number of mismatches is 2, which _is_ greater than `--min-delta`, and therefore the read is assigned to barcode `b1`.
+3. If b1 matches with 0 mismatch, and b2 matches with 2 mismatches, then the delta between the number of mismatches is 2, which _is_ greater than `--min-delta`, and therefore the read is assigned to barcode `b1`.
+4. If b1 matches with 0 mismatches, and b2 matches with 1 mismatches, then the delta between the number of mismatches is 1, which _is not_ greater than `--min-delta`, and therefore the read is not assigned to a barcode.
+5. If b1 matches with 3 mismatches, and b2 matches with 6 mismatches, then the delta between the number of mismatches is 2, which _is_ greater than `--min-delta`, and the number of mismatches for b1 is less than equal to than `--allowed-mismatches`, and thefore read is assigned to barcode `b1`.
```
Comment (nh13) [inline README.md:92]:
```suggestion
5. If b1 matches with 3 mismatches, and b2 matches with 6 mismatches, then the delta between the number of mismatches is 3, which _is_ greater than `--min-delta`, and the number of mismatches for b1 is less than equal to than `--allowed-mismatches`, and thefore read is assigned to barcode `b1`.
```

File: README.md [hunk 5]
```diff
@@ -86,12 +86,13 @@ Note: the allowed mismatches is not used when determining the next-best matching
...
+6. If b1 matches with 4 mismatches, and b2 matches with 6 mismatches, then the delta between the number of mismatches is 2, which _is_ greater than `--min-delta`, and but the number of mismatches for b1 is greater than `--allowed-mismatches`, and thefore the read is not assigned to a barcode.
```
Comment (nh13) [inline README.md:93]:
```suggestion
6. If b1 matches with 4 mismatches, and b2 matches with 6 mismatches, then the delta between the number of mismatches is 2, which _is_ greater than `--min-delta`, but the number of mismatches for b1 is greater than `--allowed-mismatches`, and thefore the read is not assigned to a barcode.
```

File: README.md [hunk 6]
```diff
@@ -86,12 +86,13 @@ Note: the allowed mismatches is not used when determining the next-best matching
...
+7. If b1 matches with 2 mismatch, and b2 matches with 2 mismatches, then the delta between the number of mismatches is 0, which _is not_ greater than `--min-delta`, and therefore the read is not assigned to a
+barcode.
```
[the remark this answers is not in the document]
Comment (theJasonFan) [reply README.md]: Noted, and reverted in 9efdf9074bf487cb33ffd15b35488401894874e4

**Comment by theJasonFan** [comment]
Noting here that the tests in `matcher.rs` could parameterized like tests in `run.rs` with parameterized tests using `rstest`. But I think what we haver right now is fine and is very explicit. I've squashed the PR into two commits one with the failing test and one with the fix. Merging!

**Pull request merged**